router/pkg/watcher/simple_watcher_test.go

```diff
@@ -0,0 +1,286 @@
+}
+
+var watchInterval = 10 * time.Millisecond
+
+func TestWatch(t *testing.T) {
+	t.Parallel()
+
+	t.Run("create and move", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		var err error
+
+		dir := t.TempDir()
+		tempFile := filepath.Join(dir, "config.json")
+
+		err = os.WriteFile(tempFile, []byte("a"), 0o600)
+		require.NoError(t, err)
+
+		wg := sync.WaitGroup{}
+
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			return watcher.SimpleWatch(ctx, watcher.SimpleWatcherOptions{
+				Interval: watchInterval,
+				Logger:   zap.NewNop(),
+				Path:     tempFile,
+				Callback: func() {
+					wg.Done()
+				},
+			})
+		})
+
+		// Wait for the first cycle to complete to set baseline
+		time.Sleep(2 * watchInterval)
+
+		wg.Add(1)
+
+		tempFile2 := filepath.Join(dir, "config2.json")
+
+		// Careful, this is subtly timing dependent. If we create
+		// the new file too quickly after the first, some filesystems
+		// will not record a different timestamp between the two files.
+		// The sleep above should be adequate, but if you're not
+		// seeing the event, try increasing it.
+		err = os.WriteFile(tempFile2, []byte("b"), 0o600)
+		require.NoError(t, err)
+
+		// Move new file ontop of the old file
+		err = os.Rename(tempFile2, tempFile)
+		require.NoError(t, err)
+
+		// Should get an event for the new file
+		wg.Wait()
+	})
+
+	t.Run("modify an existing file", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		dir := t.TempDir()
+		tempFile := filepath.Join(dir, "config.json")
+
+		err := os.WriteFile(tempFile, []byte("a"), 0o600)
+		require.NoError(t, err)
+
+		wg := sync.WaitGroup{}
+
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			return watcher.SimpleWatch(ctx, watcher.SimpleWatcherOptions{
+				Interval: watchInterval,
+				Logger:   zap.NewNop(),
+				Path:     tempFile,
+				Callback: func() {
+					wg.Done()
+				},
+			})
+		})
+
+		// Wait for the first cycle to complete to set baseline
+		time.Sleep(2 * watchInterval)
+
+		wg.Add(1)
+
+		err = os.WriteFile(tempFile, []byte("b"), 0o600)
+		require.NoError(t, err)
+
+		wg.Wait()
+	})
+
+	t.Run("delete and replace a file", func(t *testing.T) {
+		t.Parallel()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		dir := t.TempDir()
+		tempFile := filepath.Join(dir, "config.json")
+
+		err := os.WriteFile(tempFile, []byte("a"), 0o600)
+		require.NoError(t, err)
+
+		wg := sync.WaitGroup{}
+
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			return watcher.SimpleWatch(ctx, watcher.SimpleWatcherOptions{
+				Interval: watchInterval,
+				Logger:   zap.NewNop(),
+				Path:     tempFile,
+				Callback: func() {
+					wg.Done()
+				},
+			})
+		})
+
+		// Wait for the first cycle to complete to set baseline
+		time.Sleep(2 * watchInterval)
+
+		wg.Add(1)
```

Comment:
Just from a review perspective wg here looks a bit like a race condition. I personally would try to avoid time.Sleep. I suggest adding something to the waitgroup first and waiting 2 times or you could also work with channels and a select with something like `<-time.After`.

What do you think?

Reply:
Can you clarify what you think looks like a race here? 

To clarify the control flow:

- We want to ensure the callback is run when the file changes
- We set up a waitgroup, and have the callback call `wg.Done()`
- We sleep a bit to let the first watcher cycle pass, otherwise it will not detect the change as the first pass result is "baseline"
- We delete the file, and wait for another cycle, deletions do not trigger reloads
- We write the file back, and then wait on the callback waitgroup to ensure it runs

WaitGroup is sound here because:

- If it runs more than once, `wg.Done()` will panic
- If it runs less than once, the test will time out and fail
- It it runs exactly once, the test will pass

Thoughts?